crates/compiler/src/output/declaration.rs

```diff
@@ -0,0 +1,239 @@
+       /// inferred from usage. If <see langword="false"/>, this Declaration
+       /// appears in the source code.</value>
+       public bool IsImplicit { get; internal set; }
+
+       /// <summary>
+       /// Gets the types of the variable, as represented by an object that
+       /// implements <see cref="IType"/>.
+       /// </summary>
+       public Yarn.IType Type { get; internal set; }
+
+       /// <summary>
+       /// The string used for <see cref="SourceFileName"/> if the
+       /// Declaration was found outside of a Yarn source file.
+       /// </summary>
+       public const string ExternalDeclaration = "(External)";
+
+       /// <summary>
+       /// Gets the range of text at which this declaration occurs.
+       /// </summary>
+       /// <remarks>
+       /// This range refers to the declaration of the symbol itself, and not
+       /// any syntax surrounding it. For example, the declaration
+       /// <c>&lt;&lt;declare $x = 1&gt;&gt;</c> would have a Range referring
+       /// to the <c>$x</c> symbol.
+       /// </remarks>
+       public Range Range { get; internal set; } = new Range();
+*/
+
+use rusty_yarn_spinner_core::prelude::{Type, Value};
+use std::ops::RangeInclusive;
+
+/// Information about a declaration. Stored inside a declaration table,
+/// which is produced from the Compiler.
+///
+/// You do not create instances of this class yourself. They are
+/// generated by the [`Compiler`].
+#[derive(Debug, Clone)]
+pub struct Declaration {
+    /// The name of this declaration.
+    pub name: String,
+
+    /// The default value of this declaration, if no value has been
+    /// specified in code or is available from a [`Dialogue`]'s
+    /// [`IVariableStorage`].
+    pub default_value: Value,
+
+    /// A string describing the purpose of this declaration.
+    pub description: String,
+
+    /// The name of the file in which this declaration was found.
+    ///
+    /// If this declaration was not found in a Yarn source file, this
+    /// will be [`DeclarationSource::External`].
+    pub source_file_name: DeclarationSource,
+
+    /// The name of the node in which this declaration was found.
+    ///
+    /// If this declaration was not found in a Yarn source file, this
+    /// will be [`None`].
+    pub source_node_name: Option<String>,
+
+    /// The line number at which this declaration was found in the
+    /// source file.
+    ///
+    /// If this declaration was not found in a Yarn source file, this
+    /// will be [`None`].
+    pub source_file_line: Option<usize>,
+
+    /// A value indicating whether this declaration was implicitly
+    /// inferred from usage.
+    ///
+    /// If `true`, this declaration was implicitly inferred from usage.
+    /// If `false`, this declaration appears in the source code.
+    pub is_implicit: bool,
+
+    /// The types of the variable, as represented by an object found
+    /// in a variant of [`DeclarationType`].
+    pub r#type: Type, // TODO

```

Comment:
r# is to avoid naming conflicts, right? As type is a protected keyword?

Reply:
Exactly